source/extensions/quic_listeners/quiche/envoy_quic_client_session.cc

```diff
@@ -0,0 +1,78 @@
+#include "extensions/quic_listeners/quiche/envoy_quic_client_session.h"
+
+namespace Envoy {
+namespace Quic {
+
+EnvoyQuicClientSession::EnvoyQuicClientSession(
+    const quic::QuicConfig& config, const quic::ParsedQuicVersionVector& supported_versions,
+    std::unique_ptr<EnvoyQuicClientConnection> connection, const quic::QuicServerId& server_id,
+    quic::QuicCryptoClientConfig* crypto_config,
+    quic::QuicClientPushPromiseIndex* push_promise_index, Event::Dispatcher& dispatcher,
+    uint32_t send_buffer_limit)
+    : QuicFilterManagerConnectionImpl(*connection, dispatcher, send_buffer_limit),
+      quic::QuicSpdyClientSession(config, supported_versions, connection.release(), server_id,
+                                  crypto_config, push_promise_index) {}
+
+EnvoyQuicClientSession::~EnvoyQuicClientSession() {
+  ASSERT(!connection()->connected());
+  quic_connection_ = nullptr;
```

Comment:
Big +1. This is what I was going to suggest actually. If we have a parent class in which each of the Quic/Envoy classes are members, with a parent pointer back to the holding class, we could also get rid of all of the dynamic casts. Theoretically it might perform a bit worse with having to traverse the parent object (maybe) but IMO it will make everything a lot easier to understand. Glad to hear you will clean this up in a follow up.